libs/vkd3d/command.c

```diff
@@ -3474,87 +3474,231 @@ static HRESULT d3d12_command_list_allocate_transfer_buffer(struct d3d12_command_
+
+        vk_descriptor_write.sType = VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET;
+        vk_descriptor_write.pNext = NULL;
+        vk_descriptor_write.dstSet = vk_descriptor_set;
+        vk_descriptor_write.dstBinding = 0;
+        vk_descriptor_write.dstArrayElement = 0;
+        vk_descriptor_write.descriptorCount = 1;
+        vk_descriptor_write.descriptorType = VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE;
+        vk_descriptor_write.pImageInfo = &vk_image_info;
+        vk_descriptor_write.pBufferInfo = NULL;
+        vk_descriptor_write.pTexelBufferView = NULL;
+
+        VK_CALL(vkUpdateDescriptorSets(list->device->vk_device, 1, &vk_descriptor_write, 0, NULL));
+
+        for (i = 0; i < ARRAY_SIZE(vk_image_barriers); i++)
+        {
+            vk_image_barriers[i].sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER;
+            vk_image_barriers[i].pNext = NULL;
+            vk_image_barriers[i].srcAccessMask = 0;
+            vk_image_barriers[i].dstAccessMask = 0;
+            vk_image_barriers[i].srcQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+            vk_image_barriers[i].dstQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+        }
+
+        attachment_layout = dst_is_depth_stencil
+                ? VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL
+                : VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL;
 
-    assert(d3d12_resource_desc_get_width(src_desc, src_miplevel_idx) ==
-            d3d12_resource_desc_get_width(dst_desc, dst_miplevel_idx));
-    assert(d3d12_resource_desc_get_height(src_desc, src_miplevel_idx) ==
-            d3d12_resource_desc_get_height(dst_desc, dst_miplevel_idx));
-    assert(d3d12_resource_desc_get_depth(src_desc, src_miplevel_idx) ==
-            d3d12_resource_desc_get_depth(dst_desc, dst_miplevel_idx));
+        vk_image_barriers[0].oldLayout = dst_layout;
+        vk_image_barriers[0].newLayout = attachment_layout;
+        vk_image_barriers[0].image = dst_resource->u.vk_image;
+        vk_image_barriers[0].subresourceRange = vk_subresource_range_from_layers(&region->dstSubresource);
 
-    VK_CALL(vkCmdCopyBufferToImage(list->vk_command_buffer,
-            transfer_buffer.vk_buffer, dst_resource->u.vk_image,
-            VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, 1, &buffer_image_copy));
+        if (region->extent.width == extent.width && region->extent.height == extent.height)
+            vk_image_barriers[0].oldLayout = VK_IMAGE_LAYOUT_UNDEFINED;
+
+        vk_image_barriers[1].oldLayout = src_layout;
+        vk_image_barriers[1].newLayout = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
+        vk_image_barriers[1].dstAccessMask = VK_ACCESS_SHADER_READ_BIT;
+        vk_image_barriers[1].image = src_resource->u.vk_image;
+        vk_image_barriers[1].subresourceRange = vk_subresource_range_from_layers(&region->srcSubresource);
+
+        VK_CALL(vkCmdPipelineBarrier(list->vk_command_buffer,
+                VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT,
```

Comment:
Shouldn't this be TRANSFER_BIT since that's where we expected to consume a copy in the first place?